### PR TITLE
Xamarin mac support + refactoring

### PR DESCRIPTION
**Pull request by olegoid**
This PR brings the following features to xamarin-builder step:
- Xamarin.Mac projects support
- F# projects support
- New approach of project type detection: use project type guids instead of referenced libraries

### CODE REVIEW

**Comment by viktorbenei** [comment]
Thanks, looks fantastic - @birmacher will review & test this after Evolve! ;)

**Comment by birmacher** [comment]
I'll try to do the review now, wish me luck! I'm fighting with the WIFI on the plane 🤓

File: generate_export_options.rb [hunk 1]
```diff
@@ -87,16 +92,23 @@ def export_method(mobileprovision_content)
...
+mobileprovision_content = collect_provision_info(options[:archive_path], options[:target_os])
 # team_id = mobileprovision_content['TeamIdentifier'].first
 method = export_method(mobileprovision_content)
 
 
 export_options = {}
 export_options[:method] = method unless method.nil?
+
+# explicitly set this option to false for Xamarin.Mac projects since they have no support of dSYMs
+export_options[:uploadSymbols] = 'NO' if options[:target_os].eql?(Api::MAC)
```
Comment (birmacher) [inline generate_export_options.rb]: Should we let the users know about it? I mean we can give them a bit of detail so they will understand why it's skipped, like:

Warning: Skipping dSym upload; not supported by Mac OS projects

Comment (viktorbenei) [inline generate_export_options.rb]: yep, logging is quite important - we want to keep our support guys (ahem, us :P) sane ;)

File: step.yml [hunk 2]
```diff
@@ -42,7 +42,10 @@ inputs:
       value_options:
       - "ios"
       - "android"
+      - "mac"
       - "ios,android"
+      - "ios,mac"
+      - "android,mac"
```
Comment (birmacher) [inline step.yml:48]: ios,android,mac is missing from the list

**Comment by birmacher** [comment]
@olegoid Thanks for the PR! 🎉 

I've added two comments above, please check them if you have time.
I'll do a couple of local testing a bit later and after everything looks cool we can move the PR to this repository: https://github.com/bitrise-io/xamarin-builder
This is the common repo for the xamarin projects.

File: step.yml (hunk 2 above)
Comment (olegoid) [inline step.yml:48]: OK, will do.

Comment (olegoid) [inline step.yml:48]: Should be fixed in  246fd5f

File: generate_export_options.rb (hunk 1 above)
Comment (olegoid) [inline generate_export_options.rb]: Fixed in 06e696e

**Comment by godrei** [comment]
Hi @olegoid !
Thanks for the contribution!
Xamarin.Mac projects support solved in [v1.3.0](https://github.com/bitrise-steplib/steps-xamarin-builder/releases/tag/1.3.0)

**Pull request closed without merging**